### PR TITLE
GH-40036: [C++] Azure file system write buffering & async writes

### PR DESCRIPTION
**Pull request by OliLay**
### Rationale for this change

See #40036.

### What changes are included in this PR?

Write buffering and async writes (similar to what the S3 file system does) in the `ObjectAppendStream` for the Azure file system.

With write buffering and async writes, the input scenario creation runtime in the tests (which uses the `ObjectAppendStream` against Azurite) decreased from ~25s (see [here](https://github.com/apache/arrow/issues/40036)) to ~800ms:
```
[ RUN      ] TestAzuriteFileSystem.OpenInputFileMixedReadVsReadAt
[       OK ] TestAzuriteFileSystem.OpenInputFileMixedReadVsReadAt (787 ms)
```

### Are these changes tested?
Added some tests with background writes enabled and disabled (some were taken from the S3 tests). Everything changed should be covered.

### Are there any user-facing changes?
`AzureOptions` now allows for `background_writes` to be set (default: true). No breaking changes.

### Notes

- The code in `DoWrite` is very similar to [the code in the S3 FS](https://github.com/apache/arrow/blob/edfa343eeca008513f0300924380e1b187cc976b/cpp/src/arrow/filesystem/s3fs.cc#L1753). Maybe this could be unified? I didn't see this in the scope of the PR though.
* GitHub Issue: #40036

### CODE REVIEW

**Comment by OliLay** [comment]
Hi @kou, sorry for directly pinging you again, but do you maybe know who would be an appropriate person to look at this PR? Same for #43098

**Comment by kou** [comment]
No problem. Sorry for not reviewing this and https://github.com/apache/arrow/pull/43098.
@Tom-Newton @felipecrv and I implemented Azure filesystem. So one of us are appropriate person to ping.

File: cpp/src/arrow/filesystem/azurefs.cc [hunk 1]
```diff
@@ -1066,20 +1098,88 @@ class ObjectAppendStream final : public io::OutputStream {
...
-    return CommitBlockList(block_blob_client_, block_ids_, commit_block_list_options_);
+
+    auto fut = FlushAsync();
+    RETURN_NOT_OK(fut.status());
+
+    return CommitBlockList();
+  }
+
+  Future<> FlushAsync() {
+    RETURN_NOT_OK(CheckClosed("flush"));
```
Comment (kou) [inline cpp/src/arrow/filesystem/azurefs.cc:1109]: Can we move https://github.com/apache/arrow/blob/0bae073d1abe439d113cc12e06e7ada886a2f2fd/cpp/src/arrow/filesystem/azurefs.cc#L1063-L1067 from `Flush()`?
It seems that we don't need to execute `CheckClosed("flush")` and `if (!initialized_)` in both of `Flush()` and `FlushAsync()`.

File: cpp/src/arrow/filesystem/azurefs.cc [hunk 2]
```diff
@@ -1066,20 +1098,88 @@ class ObjectAppendStream final : public io::OutputStream {
...
+                 std::shared_ptr<Buffer> owned_buffer = nullptr) {
+    if (closed_) {
+      return Status::Invalid("Operation on closed stream");
+    }
+
+    const auto* data_ptr = reinterpret_cast<const int8_t*>(data);
+    auto advance_ptr = [&data_ptr, &nbytes](const int64_t offset) {
+      data_ptr += offset;
+      nbytes -= offset;
+    };
```
Comment (kou) [inline cpp/src/arrow/filesystem/azurefs.cc:1134]: How about updating `pos_` and `content_length_` too in this and improve variable name for the change?

File: cpp/src/arrow/filesystem/azurefs_test.cc [hunk 3]
```diff
@@ -1471,6 +1473,93 @@ class TestAzureFileSystem : public ::testing::Test {
...
   }
 
+  void AssertObjectContents(AzureFileSystem* fs, std::string_view path,
+                            std::string_view expected) {
+    ASSERT_OK_AND_ASSIGN(auto input, fs->OpenInputStream(std::string{path}));
+    std::string contents;
+    std::shared_ptr<Buffer> buffer;
+    do {
+      ASSERT_OK_AND_ASSIGN(buffer, input->Read(128 * 1024));
+      ASSERT_TRUE(buffer);
```
Comment (kou) [inline cpp/src/arrow/filesystem/azurefs_test.cc:1483]: Why do we need this?

File: cpp/src/arrow/filesystem/azurefs_test.cc [hunk 4]
```diff
@@ -1471,6 +1473,93 @@ class TestAzureFileSystem : public ::testing::Test {
...
+    AssertObjectContents(fs.get(), path, expected);
+  }
+
+  void TestOpenOutputStreamLarge() {
+    ASSERT_OK_AND_ASSIGN(auto fs, AzureFileSystem::Make(options_));
+
+    auto data = SetUpPreexistingData();
+    const auto path = data.ContainerPath("test-write-object");
+    ASSERT_OK_AND_ASSIGN(auto output, fs->OpenOutputStream(path, {}));
+    std::array<std::int64_t, 3> sizes{2570 * 1024, 258 * 1024, 259 * 1024};
```
Comment (kou) [inline cpp/src/arrow/filesystem/azurefs_test.cc:1510]: Could you add a comment why we should use these sizes?

File: cpp/src/arrow/filesystem/azurefs_test.cc [hunk 5]
```diff
@@ -1471,6 +1473,93 @@ class TestAzureFileSystem : public ::testing::Test {
...
+    std::array<std::string, 3> buffers{
+        std::string(sizes[0], 'A'),
+        std::string(sizes[1], 'B'),
+        std::string(sizes[2], 'C'),
+    };
+    auto expected = std::int64_t{0};
+    for (auto i = 0; i != 3; ++i) {
```
Comment (kou) [inline cpp/src/arrow/filesystem/azurefs_test.cc:1517]: Can we use `i < buffers.size()` here?

File: cpp/src/arrow/filesystem/azurefs_test.cc [hunk 6]
```diff
@@ -1471,6 +1473,93 @@ class TestAzureFileSystem : public ::testing::Test {
...
+    auto data = SetUpPreexistingData();
+    const auto path = data.ContainerPath("test-write-object");
+    ASSERT_OK_AND_ASSIGN(auto output, fs->OpenOutputStream(path, {}));
+    std::array<std::int64_t, 3> sizes{2570 * 1024, 258 * 1024, 259 * 1024};
+    std::array<std::string, 3> buffers{
+        std::string(sizes[0], 'A'),
+        std::string(sizes[1], 'B'),
+        std::string(sizes[2], 'C'),
+    };
+    auto expected = std::int64_t{0};
```
Comment (kou) [inline cpp/src/arrow/filesystem/azurefs_test.cc:1516]:
```suggestion
    auto expected_size = std::int64_t{0};
```

File: cpp/src/arrow/filesystem/azurefs_test.cc [hunk 7]
```diff
@@ -1471,6 +1473,93 @@ class TestAzureFileSystem : public ::testing::Test {
...
+  }
+
+  void TestOpenOutputStreamCloseAsyncDestructor() {
+    ASSERT_OK_AND_ASSIGN(auto fs, AzureFileSystem::Make(options_));
+    std::shared_ptr<io::OutputStream> stream;
+    auto data = SetUpPreexistingData();
+    const std::string path = data.ContainerPath("test-write-object");
+    constexpr auto payload = PreexistingData::kLoremIpsum;
+
+    ASSERT_OK_AND_ASSIGN(stream, fs->OpenOutputStream(path));
```
Comment (kou) [inline cpp/src/arrow/filesystem/azurefs_test.cc:1534]:
```suggestion
    auto data = SetUpPreexistingData();
    const std::string path = data.ContainerPath("test-write-object");
    constexpr auto payload = PreexistingData::kLoremIpsum;

    ASSERT_OK_AND_ASSIGN(auto stream, fs->OpenOutputStream(path));
```

File: cpp/src/arrow/filesystem/azurefs_test.cc [hunk 8]
```diff
@@ -1471,6 +1473,93 @@ class TestAzureFileSystem : public ::testing::Test {
...
+    ASSERT_OK(stream->Write(payload));
+    // Destructor implicitly closes stream and completes the multipart upload.
+    // GH-37670: Testing it doesn't matter whether flush is triggered asynchronously
```
Comment (kou) [inline cpp/src/arrow/filesystem/azurefs_test.cc:1537]: It seems that GH-37670 is for S3 filesystem.
Is this true for Azure filesystem too?

File: cpp/src/arrow/filesystem/azurefs_test.cc [hunk 9]
```diff
@@ -1471,6 +1473,93 @@ class TestAzureFileSystem : public ::testing::Test {
...
+  }
+
+  void TestOpenOutputStreamDestructor() {
+    ASSERT_OK_AND_ASSIGN(auto fs, AzureFileSystem::Make(options_));
+    std::shared_ptr<io::OutputStream> stream;
+    constexpr auto* payload = "new data";
+    auto data = SetUpPreexistingData();
+    const std::string path = data.ContainerPath("test-write-object");
+
+    ASSERT_OK_AND_ASSIGN(stream, fs->OpenOutputStream(path));
```
Comment (kou) [inline cpp/src/arrow/filesystem/azurefs_test.cc:1555]:
```suggestion
    constexpr auto* payload = "new data";
    auto data = SetUpPreexistingData();
    const std::string path = data.ContainerPath("test-write-object");

    ASSERT_OK_AND_ASSIGN(auto stream, fs->OpenOutputStream(path));
```

File: cpp/src/arrow/filesystem/azurefs.cc [hunk 10]
```diff
@@ -1066,20 +1098,88 @@ class ObjectAppendStream final : public io::OutputStream {
...
+      advance_ptr(kBlockUploadSize);
+      pos_ += kBlockUploadSize;
+      content_length_ += kBlockUploadSize;
+    }
+
+    // Buffer remaining bytes
+    if (nbytes > 0) {
+      current_block_size_ = nbytes;
+      ARROW_ASSIGN_OR_RAISE(current_block_, io::BufferOutputStream::Create(
+                                                kBlockUploadSize, io_context_.pool()));
```
Comment (kou) [inline cpp/src/arrow/filesystem/azurefs.cc:1167]: Can we reuse `Reset()`-ed `current_block_` instead of creating a new one?

File: cpp/src/arrow/filesystem/azurefs.h [hunk 11]
```diff
@@ -112,6 +112,9 @@ struct ARROW_EXPORT AzureOptions {
   /// This will be ignored if non-empty metadata is passed to OpenOutputStream.
   std::shared_ptr<const KeyValueMetadata> default_metadata;
 
+  /// Whether OutputStream writes will be issued in the background, without blocking.
+  bool background_writes = true;
```
Comment (kou) [inline cpp/src/arrow/filesystem/azurefs.h:116]: How about referring `ARROW_ENABLE_THREADING` here?

```suggestion
#ifdef ARROW_ENABLE_THREADING
  bool background_writes = true;
#else
  bool background_writes = false;
#endif
```

**Review by kou — COMMENTED** [review]
> The code in DoWrite is very similar to [the code in the S3 FS](https://github.com/apache/arrow/blob/edfa343eeca008513f0300924380e1b187cc976b/cpp/src/arrow/filesystem/s3fs.cc#L1753). Maybe this could be unified? I didn't see this in the scope of the PR though.

It's better if it improve maintainability. We can work on it as a follow-up task.

File: cpp/src/arrow/filesystem/azurefs.h (hunk 11 above)
Comment (felipecrv) [reply cpp/src/arrow/filesystem/azurefs.h:116]: Shouldn't this be set to `false` in the beginning before potential bugs are ironed out?

Comment (OliLay) [reply cpp/src/arrow/filesystem/azurefs.h:116]: Fine for me, lets start with disabling it per default at first.

File: cpp/src/arrow/filesystem/azurefs_test.cc (hunk 3 above)
Comment (OliLay) [reply cpp/src/arrow/filesystem/azurefs_test.cc:1483]: I guess it asserts that reading the input in fact worked (because that is also using the underlying Azure file system implementation). But we can also get rid of it, I don't have a strong opinion on it but I also don't see any harm.

File: cpp/src/arrow/filesystem/azurefs_test.cc (hunk 8 above)
Comment (OliLay) [reply cpp/src/arrow/filesystem/azurefs_test.cc:1537]: Yes, not the exact wording, as we don't use a "multipart upload" but we also trigger the upload of the last data, if any, and then commit it. So there is also some I/O when the stream is closed which we need to make sure we wait on. I'll clarified the comment and removed the ref to the GitHub issue.

File: cpp/src/arrow/filesystem/azurefs_test.cc (hunk 4 above)
Comment (OliLay) [reply cpp/src/arrow/filesystem/azurefs_test.cc:1510]: These sizes were kind of arbitrary (I did not introduce them), I changed to more reasonable sizes. The rationale of the test is just to issue different sizes of writes that trigger different mechanisms (e.g. buffering, directly uploading, etc.).

File: cpp/src/arrow/filesystem/azurefs.cc (hunk 1 above)
Comment (OliLay) [reply cpp/src/arrow/filesystem/azurefs.cc:1109]: I moved this to `Flush`, since this is the public API.

**Comment by OliLay** [comment]
> > The code in DoWrite is very similar to [the code in the S3 FS](https://github.com/apache/arrow/blob/edfa343eeca008513f0300924380e1b187cc976b/cpp/src/arrow/filesystem/s3fs.cc#L1753). Maybe this could be unified? I didn't see this in the scope of the PR though.
> 
> It's better if it improve maintainability. We can work on it as a follow-up task.

Yeah, I just have to admit that I haven't come up with a good idea to factor this out and provide a good abstraction on top. Because this functionality may be needed by any file system that implements write buffering.

File: cpp/src/arrow/filesystem/azurefs.cc [hunk 12]
```diff
@@ -1066,20 +1098,91 @@ class ObjectAppendStream final : public io::OutputStream {
       // flush. This also avoids some unhandled errors when flushing in the destructor.
       return Status::OK();
     }
-    return CommitBlockList(block_blob_client_, block_ids_, commit_block_list_options_);
+
+    auto fut = FlushAsync();
+    RETURN_NOT_OK(fut.status());
+
+    return CommitBlockList();
```
Comment (felipecrv) [inline cpp/src/arrow/filesystem/azurefs.cc:1105]: I think it's better to keep the `CommitBlockList` version that takes parameters and have all callers acquire the lock before calling it.

Comment (felipecrv) [reply cpp/src/arrow/filesystem/azurefs.cc:1105]: In general, keep lock acquisition "in the outside" as much as possible.

Comment (OliLay) [reply cpp/src/arrow/filesystem/azurefs.cc]: Addressed your feedback in [4ce4ea1](https://github.com/apache/arrow/pull/43096/commits/4ce4ea1520c90ed70b530d88b9042e1a468fa77f)

**Review by felipecrv — COMMENTED** [review]
This is a complicated change (at least for me) and I don't currently have time to give a full review.

File: cpp/src/arrow/filesystem/azurefs.cc [hunk 13]
```diff
@@ -1027,16 +1052,38 @@ class ObjectAppendStream final : public io::OutputStream {
     return Status::OK();
   }
 
+  Status EnsureReadyToFlushFromClose() {
```
[the remark this answers is not in the document]
Comment (OliLay) [reply cpp/src/arrow/filesystem/azurefs.cc:1055]: I think this is more of a style question, as this function is called twice from different places, I actually prefer it being a separate method to remove duplication. But I don't have a strong opinion, so I'll just removed it.

File: cpp/src/arrow/filesystem/azurefs.cc [hunk 14]
```diff
@@ -1066,20 +1113,95 @@ class ObjectAppendStream final : public io::OutputStream {
...
       return Status::OK();
     }
-    return CommitBlockList(block_blob_client_, block_ids_, commit_block_list_options_);
+
+    auto fut = FlushAsync();
+    RETURN_NOT_OK(fut.status());
+
+    std::unique_lock<std::mutex> lock(upload_state_->mutex);
+    return CommitBlockList(block_blob_client_, upload_state_->block_ids,
+                           commit_block_list_options_);
```
[the remark this answers is not in the document]
Comment (OliLay) [reply cpp/src/arrow/filesystem/azurefs.cc:1122]: If we restructure the lock as you've suggested, we get a deadlock because the future can actually never be marked as finished by a succeeding background write (because in `HandleUploadOutcome`, where we try to mark the future as finished, we also need to acquire the lock).
I think the race you pointed out is only an issue if there is a parallel `Write()` and `Flush()` to the stream. That is the only way how `pending_blocks_completed` could change the state during `Flush()`. To my knowledge, arrow's output streams are not supposed to be thread-safe (e.g. the S3 output stream isn't), so it should not be a problem in practice.

[the remark this answers is not in the document]
Comment (felipecrv) [reply cpp/src/arrow/filesystem/azurefs.cc:1119]: This is very important information that should be in the source code. Multi-threaded code should have comments containing arguments for its correctness.

[the remark this answers is not in the document]
Comment (OliLay) [reply cpp/src/arrow/filesystem/azurefs.cc:1120]: I added a comment to indicate that it is not thread-safe. [2ffbcb3](https://github.com/apache/arrow/pull/43096/commits/2ffbcb3e644290b3a66270392c50593d59d5bab8)

**Comment by felipecrv** [comment]
@Tom-Newton can you take a look?

**Comment by Tom-Newton** [comment]
> @Tom-Newton can you take a look?

I can try. Maybe I can look at it tomorrow evening

File: cpp/src/arrow/filesystem/azurefs.cc (hunk 1 above)
Comment (Tom-Newton) [reply cpp/src/arrow/filesystem/azurefs.cc]: I think I agree with @kou. The S3 filesystem implements `Flush()` as
```
  Status Flush() override {
    auto fut = FlushAsync();
    return fut.status();
  }
```
I think it would be nice if we could do the same because it ensures that `Flush` and `FlushAsync` have the same behaviour.

**Review by Tom-Newton — COMMENTED** [review]
Thanks for working on this @OliLay. Hopefully my review is useful - I am a C++ noob :sweat_smile:

File: cpp/src/arrow/filesystem/azurefs.cc [hunk 15]
```diff
@@ -1031,12 +1057,31 @@ class ObjectAppendStream final : public io::OutputStream {
...
+    if (closed_) {
+      return Status::OK();
+    }
+
+    if (current_block_) {
+      // Upload remaining buffer
+      RETURN_NOT_OK(WriteBuffer());
+    }
+
+    return FlushAsync();
```
[the remark this answers is not in the document]
Comment (OliLay) [reply cpp/src/arrow/filesystem/azurefs.cc]: Good catch. I added this.

File: cpp/src/arrow/filesystem/azurefs.cc [hunk 16]
```diff
@@ -1066,20 +1111,95 @@ class ObjectAppendStream final : public io::OutputStream {
       // flush. This also avoids some unhandled errors when flushing in the destructor.
       return Status::OK();
     }
-    return CommitBlockList(block_blob_client_, block_ids_, commit_block_list_options_);
+
+    auto fut = FlushAsync();
+    RETURN_NOT_OK(fut.status());
+
+    std::unique_lock<std::mutex> lock(upload_state_->mutex);
+    return CommitBlockList(block_blob_client_, upload_state_->block_ids,
```
[the remark this answers is not in the document]
Comment (OliLay) [reply cpp/src/arrow/filesystem/azurefs.cc:1130]: Indeed, this should happen for both methods. Added it.

File: cpp/src/arrow/filesystem/azurefs.cc [hunk 17]
```diff
@@ -1066,20 +1111,95 @@ class ObjectAppendStream final : public io::OutputStream {
...
+    if (closed_) {
+      return Status::Invalid("Operation on closed stream");
+    }
+
+    const auto* data_ptr = reinterpret_cast<const int8_t*>(data);
+    auto advance_ptr = [this, &data_ptr, &nbytes](const int64_t offset) {
+      data_ptr += offset;
+      nbytes -= offset;
+      pos_ += offset;
+      content_length_ += offset;
```
[the remark this answers is not in the document]
Comment (OliLay) [reply cpp/src/arrow/filesystem/azurefs.cc:1174]: No, this will not be used concurrently. In fact, I added the disclaimer that the OutputStream API is not thread-safe. Other implementations (e.g. S3) also don't guarantee any thread safety.

File: cpp/src/arrow/filesystem/azurefs.cc [hunk 18]
```diff
@@ -1066,20 +1111,95 @@ class ObjectAppendStream final : public io::OutputStream {
...
-      return Status::OK();
+  Status WriteBuffer() {
+    ARROW_ASSIGN_OR_RAISE(auto buf, current_block_->Finish());
+    current_block_.reset();
+    current_block_size_ = 0;
+    return AppendBlock(buf);
+  }
+
+  Status DoWrite(const void* data, int64_t nbytes,
+                 std::shared_ptr<Buffer> owned_buffer = nullptr) {
```
[the remark this answers is not in the document]
Comment (OliLay) [reply cpp/src/arrow/filesystem/azurefs.cc:1164]: This is used by the `Status Write(const std::shared_ptr<Buffer>& buffer)` overload (which must be overwritten).

File: cpp/src/arrow/filesystem/azurefs.cc (hunk 1 above)
Comment (OliLay) [reply cpp/src/arrow/filesystem/azurefs.cc]: I tried to unify the sync/async flush implementations, unfortunately that did not work out due to lifetime issues in the async case. When an `ObjectAppendStream` is deconstructed in RAII way, `Close()` (and therefore `Flush()`) is called. If we call `FlushAsync()` in the close, we need to create a `shared_ptr` from `this` (to ensure lifetime of `this` when the lambda is actually called), but: we can not create a `shared_ptr` of `this` while it is deconstructed. Hence in the `Close()` call we always must do a sync `Flush` where we do not have to give these lifetime guarantees. 
TLDR: `Flush()` and `FlushAsync()` impls are similar, but slightly different and decoupled implementations.

**Comment by OliLay** [comment]
> Thanks for working on this @OliLay. Hopefully my review is useful - I am a C++ noob 😅

Thanks for your review, it is much appreciated :+1:

File: cpp/src/arrow/filesystem/azurefs.cc (hunk 16 above)
[the remark this answers is not in the document]
Comment (Tom-Newton) [reply cpp/src/arrow/filesystem/azurefs.cc:1130]: Thanks for adding it. Do the tests cover this?

File: cpp/src/arrow/filesystem/azurefs.cc (hunk 18 above)
[the remark this answers is not in the document]
Comment (Tom-Newton) [reply cpp/src/arrow/filesystem/azurefs.cc:1164]: True, but I'm a bit confused why its like this because the buffer is never read or written to as far as I can tell. It looks like S3 is the same though so I won't worry about it.

**Review by Tom-Newton — APPROVED** [review]
I'm happy :slightly_smiling_face: 

BTW one of my colleagues did a little bit of bench-marking and this change brought performance from 245 seconds to 26 seconds :rocket:

File: cpp/src/arrow/filesystem/azurefs.cc (hunk 16 above)
[the remark this answers is not in the document]
Comment (OliLay) [reply cpp/src/arrow/filesystem/azurefs.cc:1130]: Yes, this is on the code path that is tested by https://github.com/OliLay/arrow/blob/afa111bc50e2f0ec54c7eb2bdcd6982c155b75a1/cpp/src/arrow/filesystem/azurefs_test.cc#L1529

[the remark this answers is not in the document]
Comment (Tom-Newton) [reply cpp/src/arrow/filesystem/azurefs.cc:1130]: Ok, I ask because it seems the tests didn't catch this? and I happened to spot it.

[the remark this answers is not in the document]
Comment (OliLay) [reply cpp/src/arrow/filesystem/azurefs.cc:1130]: Good point. I think because in the linked test we make sure the destructor is called which will also trigger a commit. 
I wrote an more explicit test which only triggers `CloseAsync()` and makes sure data is written correctly: 48bc6bed5398f062e2f6f04b64873da7186f5f0c

File: cpp/src/arrow/filesystem/azurefs.h [hunk 19]
```diff
@@ -112,6 +112,9 @@ struct ARROW_EXPORT AzureOptions {
   /// This will be ignored if non-empty metadata is passed to OpenOutputStream.
   std::shared_ptr<const KeyValueMetadata> default_metadata;
 
+  /// Whether OutputStream writes will be issued in the background, without blocking.
+  bool background_writes = false;
```
[the remark this answers is not in the document]
Comment (Tom-Newton) [reply cpp/src/arrow/filesystem/azurefs.h:116]: I don't feel strongly, but personally I think we should stop adding functionality and perhaps remove some from  `ExtractFromUriQuery()` as was suggested in https://github.com/apache/arrow/issues/43197

[the remark this answers is not in the document]
Comment (kou) [reply cpp/src/arrow/filesystem/azurefs.h:116]: If we can specify an option value by URI something like `abfs://.../...?background_writes=yes`, we don't need to implement `AzureOptions` bindings. For example, Apache Arrow C GLib doesn't provide `S3Options` nor `GCSOptions` but it can use `S3FileSystem` and `GCSFileSystem`.
I think that #43197 is for security. But this is not related to security.

**Comment by kou** [comment]
@pitrou Could you review this?

File: cpp/src/arrow/filesystem/azurefs.cc [hunk 20]
```diff
@@ -1066,20 +1121,114 @@ class ObjectAppendStream final : public io::OutputStream {
...
+
+      RETURN_NOT_OK(current_block_->Write(data_ptr, current_block_size_));
+      pos_ += current_block_size_;
+      content_length_ += current_block_size_;
+    }
+
+    return Status::OK();
+  }
+
+  Status AppendBlock(std::shared_ptr<Buffer> buffer) {
```
Comment (pitrou) [inline cpp/src/arrow/filesystem/azurefs.cc:1225]: Can you move this near the other `AppendBlock` overload?

**Review by pitrou — CHANGES_REQUESTED** [review]
I'm not knowledgeable about Azure, but this looks reasonable in general.

File: cpp/src/arrow/filesystem/azurefs.cc [hunk 21]
```diff
@@ -1066,20 +1121,114 @@ class ObjectAppendStream final : public io::OutputStream {
...
     }
-    return CommitBlockList(block_blob_client_, block_ids_, commit_block_list_options_);
+
+    Future<> pending_blocks_completed;
+    {
+      std::unique_lock<std::mutex> lock(upload_state_->mutex);
+      pending_blocks_completed = upload_state_->pending_blocks_completed;
+    }
+
+    RETURN_NOT_OK(pending_blocks_completed.status());
```
[the remark this answers is not in the document]
Comment (OliLay) [reply cpp/src/arrow/filesystem/azurefs.cc:1131]: Because we can not hold the lock while waiting for the future. Another instance will acquire the lock and mark the future as completed while we wait on the future.

File: cpp/src/arrow/filesystem/azurefs.cc [hunk 22]
```diff
@@ -1066,20 +1121,114 @@ class ObjectAppendStream final : public io::OutputStream {
...
+        return Status::OK();
+      }
+
+      // Upload current buffer
+      RETURN_NOT_OK(WriteBuffer());
+    }
+
+    // We can upload chunks without copying them into a buffer
+    while (nbytes >= kBlockUploadSizeBytes) {
+      RETURN_NOT_OK(AppendBlock(data_ptr, kBlockUploadSizeBytes));
```
[the remark this answers is not in the document]
Comment (OliLay) [reply cpp/src/arrow/filesystem/azurefs.cc:1200]: What do you mean? Are you proposing to append more than `kBlockUploadSizeBytes`?

[the remark this answers is not in the document]
Comment (pitrou) [reply cpp/src/arrow/filesystem/azurefs.cc:1200]: Yes. Is it suboptimal?

[the remark this answers is not in the document]
Comment (OliLay) [reply cpp/src/arrow/filesystem/azurefs.cc]: Good point. We can make an optimization here by writing at max. 4GB (this is the blob limit for bytes per block, see [docs](https://learn.microsoft.com/en-us/azure/storage/blobs/scalability-targets)).
[57c20e9](https://github.com/apache/arrow/pull/43096/commits/57c20e99c4ba5d5c5e8f5ef8da34decf6daa9023)

File: cpp/src/arrow/filesystem/azurefs.cc (hunk 21 above)
[the remark this answers is not in the document]
Comment (pitrou) [reply cpp/src/arrow/filesystem/azurefs.cc:1133]: That makes sense indeed. Thank you!

**Review by pitrou — CHANGES_REQUESTED** [review]
Thanks a for the update @OliLay ! There's one issue that seems to need addressing (and a small nit), see below.

File: cpp/src/arrow/filesystem/azurefs.cc [hunk 23]
```diff
@@ -1066,20 +1123,110 @@ class ObjectAppendStream final : public io::OutputStream {
...
+
+      // Upload current buffer
+      RETURN_NOT_OK(AppendCurrentBlock());
+    }
+
+    // We can upload chunks without copying them into a buffer
+    while (nbytes >= kMaxBlockSizeBytes) {
+      RETURN_NOT_OK(AppendBlock(data_ptr, kMaxBlockSizeBytes));
+      advance_ptr(kMaxBlockSizeBytes);
+    }
```
[the remark this answers is not in the document]
Comment (OliLay) [reply cpp/src/arrow/filesystem/azurefs.cc]: Definitely, that was a bit too naive from my side. I think your suggestion nicely fixes this. [aea75ff](https://github.com/apache/arrow/pull/43096/commits/aea75ffcc38b484b9d960939d299b82f17b6b9d7)

**Comment by pitrou** [comment]
A small (probably innocuous) leak in reliably reported by the ASAN job. I am narrowing it down and trying to come up with a workaround:
```

==13823==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 1936 byte(s) in 2 object(s) allocated from:
    #0 0x565171d11708 in calloc (/build/cpp/debug/arrow-azurefs-test+0x4c8708) (BuildId: 308c17e28257859d0082a84d74ef009b3a4122a8)
    #0 0x7f72301e95a4 in
```

**Comment by pitrou** [comment]
A more detailed leak trace (obtained using `LSAN_OPTIONS=fast_unwind_on_malloc=0:malloc_context_size=100`) is the following:
```
Direct leak of 968 byte(s) in 1 object(s) allocated from:
    #0 0x5600b66641e8 in calloc (/build/cpp/debug/arrow-azurefs-test+0x4c31e8) (BuildId: 5f129cb231bb2651f5fb432f4c8d3c2f40506da3)
    #1 0x7f85322b25a4  (/lib/x86_64-linux-gnu/libxml2.so.2+0xe25a4) (BuildId: aebf8e42966c3ce475ff9d9d51a762831adcbb61)
    #2 0x7f85322a5fb4 in __xmlDefaultBufferSize (/lib/x86_64-linux-gnu/libxml2.so.2+0xd5fb4) (BuildId: aebf8e42966c3ce475ff9d9d51a762831adcbb61)
    #3 0x7f8532240d37 in xmlBufferCreate (/lib/x86_64-linux-gnu/libxml2.so.2+0x70d37) (BuildId: aebf8e42966c3ce475ff9d9d51a762831adcbb61)
    #4 0x7f8547ec6f74 in Azure::Storage::_internal::XmlWriter::XmlWriter() /build/cpp/_deps/azure_sdk-src/sdk/storage/azure-storage-common/src/xml_wrapper.cpp:532:19
    #5 0x7f8547e87621 in Azure::Storage::Blobs::_detail::BlockBlobClient::CommitBlockList(Azure::Core::Http::_internal::HttpPipeline&, Azure::Core::Url const&, Azure::Storage::Blobs::_detail::BlockBlobClient::CommitBlockBlobBlockListOptions const&, Azure::Core::Context const&) /build/cpp/_deps/azure_sdk-src/sdk/storage/azure-storage-blobs/src/rest_client.cpp:7944:30
    #6 0x7f8547de3149 in Azure::Storage::Blobs::BlockBlobClient::CommitBlockList(std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, Azure::Storage::Blobs::CommitBlockListOptions const&, Azure::Core::Context const&) const /build/cpp/_deps/azure_sdk-src/sdk/storage/azure-storage-blobs/src/block_blob_client.cpp:512:12
    #7 0x7f854618c14c in arrow::fs::(anonymous namespace)::CommitBlockList(std::shared_ptr<Azure::Storage::Blobs::BlockBlobClient>, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, Azure::Storage::Blobs::CommitBlockListOptions const&) /arrow/cpp/src/arrow/filesystem/azurefs.cc:948:24
    #8 0x7f8546193aab in arrow::fs::(anonymous namespace)::ObjectAppendStream::FlushAsync()::'lambda'()::operator()() const /arrow/cpp/src/arrow/filesystem/azurefs.cc:1158:14
    #9 0x7f85461933eb in std::enable_if<((!(std::is_void<arrow::Status>::value)) && (!(is_future<arrow::Status>::value))) && ((!(arrow::Future<arrow::internal::Empty>::is_empty)) || (std::is_same<arrow::Status, arrow::Status>::value)), void>::type arrow::detail::ContinueFuture::operator()<arrow::fs::(anonymous namespace)::ObjectAppendStream::FlushAsync()::'lambda'(), arrow::Status, arrow::Future<arrow::internal::Empty> >(arrow::Future<arrow::internal::Empty>, arrow::fs::(anonymous namespace)::ObjectAppendStream::FlushAsync()::'lambda'()&&) const /arrow/cpp/src/arrow/util/future.h:150:23
```

**Comment by pitrou** [comment]
The leak manifests in `TestAzuriteFileSystem.OpenOutputStreamCloseAsync` and `TestAzuriteFileSystem.OpenOutputStreamAsyncDestructor`. However, `TestAzuriteFileSystem.OpenOutputStreamAsyncDestructorNoBackgroundWrites` is fine.

**Comment by OliLay** [comment]
Interesting, I don't really see where this could come from actually. From the tests that are affected, for me it looks like that only the path when we go through `CloseAsync()/FlushAsync()` is the issue. The sync and async implementations behave very similar so I don't really see where this could come from, especially since we don't really control the Azure allocations in any way (seeing that the `XmlBuffer` is apparently the problematic allocation?)
In fact, everything Azure related should be deallocated at latest in `CloseAsync()` when the  `BlockBlobClient` should get destructed because we remove the shared ptr ref.

**Comment by pitrou** [comment]
@OliLay There's already a skip related to a libxml2 leak with threaded operation, so I added another one.

This might actually be fixed by https://github.com/Azure/azure-sdk-for-cpp/pull/5767 which was recently merged in Azure C++.

Also cc @felipecrv

**Comment by pitrou** [comment]
Actually, no, we may need something like this hack from ClickHouse:
https://github.com/ClickHouse/ClickHouse/blob/054b38d4ebeabc33010fb532766f937af2d44c03/programs/server/Server.cpp#L987-L998

It was added in this PR: https://github.com/ClickHouse/ClickHouse/pull/45796

**Comment by pitrou** [comment]
@github-actions crossbow submit -g cpp

**Comment by pitrou** [comment]
Hmm, I rebased for CI.

**Comment by pitrou** [comment]
@github-actions crossbow submit -g cpp

**Review by pitrou — APPROVED** [review]

**Comment by pitrou** [comment]
macOS build failure is unrelated as it happened elsewhere: https://github.com/apache/arrow/actions/runs/10481257647/job/29030430952
